### DISCUSSION TITLE
Change PDF version

### DISCUSSION THREAD

**Discussion by Ectalite** [tectonic-typesetting/tectonic#1192] [Q&A]
I'm including a PDF page inside my Tex project and since that I'm getting the warning:
`warning: Trying to include PDF file with version (1.7), which is newer than current output PDF setting (1.5).`

After a couple of researches, you can [give a parameter when calling xelatex](https://tex.stackexchange.com/a/8826) to change the PDF version.
`xelatex -output-driver="xdvipdfmx -V 5" <yourfile>`

Is there a way to do this tectonic ?

I also tried to change `dvipdfmx.cfg` in my texlive installation, but it didn't have any impact.

**Comment by kpym** [comment]
The simplest approach might be to modify the PDF version directly from the TeX source rather than relying on compiler-specific parameters. By doing so, your code remains independent of the compiler used (whether it's Tectonic, XeLaTeX, etc.). One way to achieve this is by specifying the desired PDF version as a parameter of the `hyperref` package, like so:

```latex
\usepackage[pdfversion=1.7]{hyperref}
```

Additional solutions can be explored in the responses to this [TeX.SX question](https://tex.stackexchange.com/q/617970).

**Reply by Ectalite** [reply]
I tried this too, but it doesn't to work for my use case. Maybe another package is changing that...
You can find my preamble here:
[preambel.zip](https://github.com/tectonic-typesetting/tectonic/files/15307476/preambel.zip)

EDIT: I also tried the solutions from this [Tex.SX question](https://tex.stackexchange.com/questions/638194/how-to-generate-pdf-1-7-output-with-lualatex) without success.

**Reply by kpym** [reply]
When I compile your preamble (with `xelatex` and with `tectonic`, after adding `\begin{document}...\end{document}`) I get `pdf 2.0` as a result:
[preambel-xelatex.pdf](https://github.com/tectonic-typesetting/tectonic/files/15307941/preambel-xelatex.pdf)

**Reply by kpym** [reply]
**Remark:** normaly `hyperref` [should be the last imported package](https://tex.stackexchange.com/q/1863), and this is not the case in your preamble.

**Reply by Ectalite** [reply]
After verification, I'm also getting a PDF with the version 2.0 in output.
But I'm also still getting the warning, so something is off.

I included the PDF with these commands:
```latex
\begin{figure}
	\centering
	\begin{tabular}{@{}c@{\hspace{.5cm}}c@{}}
		\includegraphics[page=7, width=.45\textwidth]{anhang/PDF.pdf} &
		\includegraphics[page=8, width=.45\textwidth]{anhang/PDF.pdf} \\[.5cm]
	\end{tabular}
	\caption{caption}
	\label{appendix}
\end{figure}
```
If it can help, the PDF in question has the version 1.7 and was generated using Office 365.

**Reply by kpym** [reply]
I see the tectonic warning too... I have no idea what's going on here.
There is no warning when I compile with `xelatex`.

**Reply by Ectalite** [reply]
I'm gonna open an issue then, maybe someone does now about this.

**Reply by vlasakm** [reply]
[Hyperref documentation](http://mirrors.ctan.org/macros/latex/contrib/hyperref/doc/hyperref-doc.pdf) says that the option lets hyperref itself know roughly what it is allowed to do, and while it would be nice to actually make it set the PDF version, it doesn't, because there is no general way to do it.

I highlight possibilities with dvipdfmx (and by extension, Tectonic) in my answer.

**Reply by vlasakm** [reply]
```tex
\pdf_version_gset:n{1.7}
```

Seems to work:

```tex
\documentclass{article}
\usepackage{expl3}
\ExplSyntaxOn
\pdf_version_gset:n{1.7}
\ExplSyntaxOff

\begin{document}

hello world

\end{document}
```

 I checked, and actually behind the scenes they essentially do:

```tex
\special{pdf:majorversion 1}
\special{pdf:minorversion 7}
```

**Comment by vlasakm** [comment]
The warning is emitted by the "thing" actually creating the PDF, which in the case of Tectonic is (our version of) the (x)dvipdfmx tool, just like with normal XeTeX. We have to persuade it to change the PDF version it thinks it is generating.

From [it's documentation](http://mirrors.ctan.org/systems/doc/dvipdfmx/dvipdfmx.pdf) we can learn that this can be done through a "special" TeX command / primitive:

```tex
\special{pdf:minorversion 7}
```

or by specifying the `-V` command line flag with the desired version number - the documentation doesn't say this clearly, but it seems they originally assumed PDF will always be 1.X, so it is either `-V X` to set version `1.X` or `-V 2.0` to set 2.0, or maybe even `-V 1.X` works for newever versions. Though Tectonic is based on older version of dvipdfmx, and in some cases it is deliberately more limited for improving reproducibility, so while normal dvipdfmx supports passing flags either in the `dvipdfmx.conf` config file:

```
V 7
```

or by also including them directly in the document source file with `\special`:

```tex
\special{dvipdfmx:config V 7}
```

Tectonic won't accept it with `\special`. I was the one bringing support of `dvipdfmx:config` [back to Tectonic](https://github.com/tectonic-typesetting/tectonic/pull/953/commits/5f3abadfae33730ddccf052d725dd6e9c2b05357), though I didn't forsee the usefulness of `V`, so I didn't implement it, but it would make a reasonable addition IMO.

So I guess both the easiest and only option for you is:

```tex
\special{pdf:minorversion 7}
```

which seems to work.  Note that it has to be _very_ early in the document. If you have problems with it, I would appreciata a [Minimal Working Example](https://tex.meta.stackexchange.com/questions/228/ive-just-been-asked-to-write-a-minimal-working-example-mwe-what-is-that) to experiment on.

**Reply by Ectalite** [reply]
Thanks for your reply.
I tried your solutions, but I still get the warning.

I created a PDF for the minimal working example to reproduce the one I'm using for my thesis.
It's a Word document exported with the PDF-A option enabled (as PDF-A is supported from PDF version 1.7).

You can find the example here, I put every known solution until now inside.
[TectonicMinimalWorkingExample.zip](https://github.com/tectonic-typesetting/tectonic/files/15317560/TectonicMinimalWorkingExample.zip)

Edit: PDF-A option doesn't change anything, Word exports automatically to version 1.7

**Reply by vlasakm** [reply]
TL;DR: If you see just one warning about the PDF version (which you should with the settings discussed above), it should be fine. The warning is a result of weird interaction in Tectonic internals.

Thank you very much for the MWE.

I must apologize, first sorry for the delay, time has been not on my side recently. Also it seems that I misunderstood how Tectonic invokes dvipdfmx.

Without any provisions there are two warnings (only relevant log lines shown):

```console
$ tectonic thesis.tex
Running TeX ...
warning: Trying to include PDF file with version (1.7), which is newer than current output PDF setting (1.5).
Running xdvipdfmx ...
warning: Trying to include PDF file with version (1.7), which is newer than current output PDF setting (1.5).
Writing `thesis.pdf` (12.53 KiB)
```

I didn't try all, but I guess with all the options to change the PDF version of the document we can get this:

```console
$ tectonic thesis.tex
Running TeX ...
warning: Trying to include PDF file with version (1.7), which is newer than current output PDF setting (1.5).
Running xdvipdfmx ...
Writing `thesis.pdf` (12.53 KiB)
```

I.e. xdvipdfmx has the version number set correctly. But TeX itself is using dvipdfmx's code to load the PDF even before the PDF creation starts.

This is because of the unfortunate and weird model of XeTeX (and original TeX), where TeX itself doesn't really know anything about stuff like fonts, colors or images, it only has idea about rectangles with dimensions ("boxes") and spaces ("glue"), anything else is practically left to a post processor - like dvipdfmx. But in this model, TeX still needs to know the dimensions of things, to produce correctly typeset result. In your example, there is a PDF included, so e.g. to emit the message that the page is too big to be put into the current page, something has to inform TeX what are the dimensions of the PDF.

I believe that in traditional TeX and I think even in some cases for XeTeX, the `extractbb` program is used to get "bounding box" (i.e. size) of images (JPG, PNG, ...) and PDF files. Tectonic seems to load the file PDF directly to extract its dimensions (as it seems be going through the `\XeTeXpicfile` primitive, where XeTeX is not that oblivios to what it is doing). However, it uses the same code path that dvipdfmx uses normally, where usually the PDF version number is expected to be already sensibly set. But as essentially nothing from dvipdfmx has run yet, a default version number (1.5) is seen, producing a warning, which won't be produced by the following dvipdfmx run.

All in all, after you use one of the ways we discussed, and you see that the generated PDF file has the right version (it's binary, but you can just check the top which  should in ASCII say `%PDF-1.7`), and see no warning by dvipdfmx, your document is entirely correct, and the other warning is just a bug in Tectonic.

If you however see a warning by dvipdfmx, either you are using one of the ways to set the PDF version too late (and the document is likely still correct), or you are not setting the PDF version correctly, which can be problematic with some PDF processing software.

Aside: I think most software practically ignores PDF versions (at most they produce warnings like we see here), a features is either supported in the software or not, not always according to version numbers.
